[Fix #53683] Reduce cache time for non-asset files in public dir

thisismydesign opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), the motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

This Pull Request is a fix for https://github.com/rails/rails/issues/53683

### Detail

This Pull Request changes the default caching to only cache digest stamped assets for a 1 year, other public files for 1 hour only.

This is a bad default and should be fixed asap. If this is not the way, consider removing the cache altogether.

### Additional information

<!-- Provide additional information such as benchmarks, references to other repositories, or alternative solutions. -->

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

thisismydesign committed:
[Fix #53683] Reduce cache time for non-asset files in public dir

byroot commented on railties/lib/rails/generators/rails/app/templates/config/environments/production.rb.tt:
```diff
@@ -17,8 +17,17 @@ Rails.application.configure do
+  # Short cache for others: robots.txt, sitemap.xml, 404.html, etc.
+  config.public_file_server.headers = {
+    'cache-control' => lambda do |path, _|
+      if path.match?(/\A\/assets\/.*-[a-f0-9]{8,}[a-f0-9.]*$/)
```
What's the use case for downloading it?

byroot committed:
Improve default public file server cache-control further

Set `immutable` for digested assets, this improve modern browsers
behavior, saves them from ever revalidating.

Set `stale-while-revalidate` for non-digested files. This allows
caching proxies to ever only revalidate at most once per minute
and asyncrhonously.

jdelStrother commented:
If it's not a dumb question... this change does work for you, right?
I'm not aware of being able to set Rack headers to procs, and when I try it I get response headers looking like this - 

```
< HTTP/1.1 200 OK
< last-modified: Wed, 18 Aug 2021 12:07:09 GMT
< content-type: text/plain
< cache-control: #<Proc:0x0000000151174718 /Users/jon/Developer/web/config/environments/development.rb:36 (lambda)>
< Content-Length: 14
```

if it works for you I'll dig into it some more, maybe I've got something weird going on locally.

byroot commented:
@jdelStrother please open a new issue. I suspect it is an issue on your side, but if it isn’t it’s a pretty bad bug so this shouldn’t be overlooked.

jdelStrother commented:
@byroot ok, done - https://github.com/rails/rails/issues/55617
